src/Pate/Memory/MemTrace.hs

```diff
@@ -1287,10 +1332,132 @@ instance OrdF (SymExpr sym) => Ord (MemFootprint sym ptrW) where
+  IsExprBuilder sym =>
+  OrdF (SymExpr sym) =>
+  sym ->
+  Map (MemFootprint sym ptrW) (Pred sym) ->
+  Map (MemFootprint sym ptrW) (Pred sym) ->
+  IO (Map (MemFootprint sym ptrW) (Pred sym))
+unionFootprintMap sym =
+  Map.mergeA
+    Map.preserveMissing
+    Map.preserveMissing
+    (Map.zipWithAMatched (\_k p1 p2 -> orPred sym p1 p2))
+
+muxFootprintMap ::
+  IsExprBuilder sym =>
+  OrdF (SymExpr sym) =>
+  sym ->
+  Pred sym ->
+  Map (MemFootprint sym ptrW) (Pred sym) ->
+  Map (MemFootprint sym ptrW) (Pred sym) ->
+  IO (Map (MemFootprint sym ptrW) (Pred sym))
+muxFootprintMap sym p =
+  Map.mergeA
+    (Map.traverseMissing (\_k x -> andPred sym x p))
+    (Map.traverseMissing (\_k y -> andPred sym y =<< notPred sym p))
+    (Map.zipWithAMatched (\_k x y -> itePred sym p x y))
+
+traceFootprintMap ::
+  IsExprBuilder sym =>
+  OrdF (SymExpr sym) =>
+  sym ->
+  MemTraceSeq sym ptrW ->
+  IO (Const (Map (MemFootprint sym ptrW) (Pred sym)) (MemEvent sym ptrW))
+traceFootprintMap sym =
+  evalWithFreshCache $ \rec -> \case
+    SymSequenceNil -> return (Const mempty)
+
+    SymSequenceCons _ (MemOpEvent x) xs ->
+      do let (fp,p) = memOpFootprint sym x
+         let m1 = Map.insert fp p mempty
+         Const m2 <- rec xs
+         Const <$> unionFootprintMap sym m1 m2
+    SymSequenceCons _ _ xs -> rec xs
+
+    SymSequenceAppend _ xs ys ->
+      do Const m1 <- rec xs
+         Const m2 <- rec ys
+         Const <$> unionFootprintMap sym m1 m2
+
+    SymSequenceMerge _ p xs ys ->
+      do Const m1 <- rec xs
+         Const m2 <- rec ys
+         Const <$> muxFootprintMap sym p m1 m2
+
+
+-- | Filter the memory event traces to leave just the observable
+--   events. For the moment, this simply discards all the
+--   @MemOp@ nodes and retains all the system call nodes.
+--
+--   We may make this more nuanced in the future by, e.g.,
+--   considering some memory operations observable (memory mapped I/O)
+--   or by considering some system calls nonobservable.
+observableEvents ::
+  IsExprBuilder sym =>
+  OrdF (SymExpr sym) =>
+  sym ->
+  MemTraceImpl sym ptrW ->
+  IO (SymSequence sym (MemEvent sym ptrW))
+observableEvents sym mem = evalWithFreshCache f (memSeq mem)
+  where
+   -- filtering function
+   flt MemOpEvent{}   = False
+   flt SyscallEvent{} = True
+
+   f _rec SymSequenceNil = nilSymSequence sym
+
+   f rec (SymSequenceCons _ x xs)
+     | flt x     = consSymSequence sym x =<< rec xs
+     | otherwise = rec xs
+
+   f rec (SymSequenceAppend _ xs ys) =
+     do xs' <- rec xs
+        ys' <- rec ys
+        appendSymSequence sym xs' ys'
+
+   f rec (SymSequenceMerge _ p xs ys) =
+     do xs' <- rec xs
+        ys' <- rec ys
+        muxSymSequence sym p xs' ys'
+
+
+-- | Get an unordered collection of all the read memory operations
+--   that occurred in this trace memory. Control-flow conditions are
+--   dropped while constructing this set, so the conditions included
+--   on these `MemOp` values may not be accurate.
```

Comment:
We could, but it's a little tricky because of how it interacts with the cache. Currently the only use site for this ignores the conditions, so it seems like a waste.  I would have preferred to move the entire `validConcreteReads` function inside this module, but that creates and awkward module import cycle.

I'll comment the situation more throughly.